lib/functions.php

```diff
@@ -812,4 +810,300 @@ function toArray(Iterator $iterator): Promise
+     */
+    function merge(array $streams): Stream
+    {
+        $source = new StreamSource;
+        $result = $source->stream();
+
+        $coroutine = coroutine(static function (Stream $stream) use (&$source) {
+            while ((null !== $value = yield $stream->continue()) && $source !== null) {
+                yield $source->emit($value);
+            }
+        });
+
+        $coroutines = [];
+        foreach ($streams as $stream) {
+            if (!$stream instanceof Stream) {
+                throw createTypeError([Stream::class], $stream);
+            }
+
+            $coroutines[] = $coroutine($stream);
+        }
+
+        Promise\all($coroutines)->onResolve(static function ($exception) use (&$source) {
+            $temp = $source;
+            $source = null;
+
+            if ($exception) {
+                $temp->fail($exception);
+            } else {
+                $temp->complete();
+            }
+        });
+
+        return $result;
+    }
+
+    /**
+     * Concatenates the given streams into a single stream, emitting from a single stream at a time. The
+     * prior stream must complete before values are emitted from any subsequent streams. Streams are concatenated
+     * in the order given (iteration order of the array).
+     *
+     * @param Stream[] $streams
+     *
+     * @return Stream
+     */
+    function concat(array $streams): Stream
+    {
+        foreach ($streams as $stream) {
+            if (!$stream instanceof Stream) {
+                throw createTypeError([Stream::class], $stream);
+            }
+        }
+
+        $source = new StreamSource;
+        $previous = [];
+        $promise = Promise\all($previous);
+
+        $coroutine = coroutine(static function (Stream $stream, callable $yield) {
+            while (null !== $value = yield $stream->continue()) {
+                yield $yield($value);
+            }
+        });
+
+        foreach ($streams as $iterator) {
+            $emit = coroutine(static function ($value) use ($source, $promise) {
+                static $pending = true, $failed = false;
+
+                if ($failed) {
+                    return;
+                }
+
+                if ($pending) {
+                    try {
+                        yield $promise;
+                        $pending = false;
+                    } catch (\Throwable $exception) {
+                        $failed = true;
+                        return; // Prior iterator failed.
+                    }
+                }
+
+                yield $source->emit($value);
+            });
+            $previous[] = $coroutine($iterator, $emit);
+            $promise = Promise\all($previous);
+        }
+
+        $promise->onResolve(static function ($exception) use ($source) {
+            if ($exception) {
+                $source->fail($exception);
+                return;
+            }
+
+            $source->complete();
+        });
+
+        return $source->stream();
```

Comment:
This implementation feels overly complicated and is probably much easier to implement using `AsyncGenerator` and a nested loop over all streams.

Reply:
I copied/pasted from the `Iterator` implementation. Wow, I have no idea why the implementation was so complex. I'll change the one for `Iterator` too.